lib/python/rosie/ws.py

```diff
@@ -189,33 +226,229 @@ def _render(self, all_revs=0, data=None, filters=None, s=None):
             filters=filters,
             s=s,
             data=data)
+        )
 
 
-def _query_parse_string(q_str):
-    """Split a query filter string into component parts."""
-    conjunction, tail = q_str.split(" ", 1)
-    if conjunction == "or" or conjunction == "and":
-        q_str = tail
-    else:
-        conjunction = "and"
-    filt = [conjunction]
-    if all(s == "(" for s in q_str.split(" ", 1)[0]):
-        start_group, q_str = q_str.split(" ", 1)
-        filt.append(start_group)
-    key, operator, value = q_str.split(" ", 2)
-    filt.extend([key, operator])
-    last_groups = value.rsplit(" ", 1)
-    if (len(last_groups) > 1 and last_groups[1] and
-            all([s == ")" for s in last_groups[1]])):
-        filt.extend(last_groups)
+class GetHandler(RosieDiscoService):
+
+    """Write out basic data for the names of standard fields or operators."""
+
+    QUERY_KEYS = [
+        "known_keys",  # Return the names of the common fields.
+        "query_operators",  # Return the allowed query operators.
+        "optional_keys",  # Return the names of the optional fields.
+    ]
+
+    def get(self, get_arg=None):
+        """Return data for basic API points of query keys without values."""
+        if (get_arg and
+                self.get_query_argument("format", default=None) == "json"):
+            for query in self.QUERY_KEYS:
+                if get_arg.startswith(query):
+                    # No need to catch AttributeError as all QUERY_KEYS valid.
+                    self.write(json.dumps(getattr(self.dao, "get_" + query)()))
+
+
+class HelloHandler(RosieDiscoService):
+
+    """Writes a 'Hello' message to the current logged-in user, else 'user'."""
+
+    HELLO = "Hello %s\n"
+
+    def get(self):
+        """Say Hello on success."""
+        data = self.HELLO % pwd.getpwuid(os.getuid()).pw_name
+        if self.get_query_argument("format", default=None) == "json":
+            self.write(json.dumps(data))
+        else:
+            self.write(data)
+
+
+class SearchHandler(RosieDiscoService):
+
+    """Serves a search of the database on the page of a given prefix."""
+
+    def get(self, _):
+        """Search database for rows with data matching the search string."""
+        s_arg = self.get_query_argument("s", default=None)
+        all_revs = self.get_query_argument("all_revs", default=0)
```

Comment:
Yes, I was aware of the clunky nature of the current handling of query parameters, as registered (fair enough if you missed this, as it is hidden away in a substantial comment):

> Clunky retrieval of query parameters: these get picked up via the regular expressions in the Tornado application handler set-up & passed as into the corresponding handler class as arguments to the get() method, but I am often using self.get_query_argument(QUERY_KEY, default=DEFAULT). We should make use of the former rather than the latter in most or even all cases, as it is cleaner & does not have to repeatedly check the URL path.

I seem to recall adding in the ``_`` parameter to ``SearchHandler.get`` simply to appease Codacy somewhat, as, while in the Tornado Handler context I think it is not necessary, it is requesting the same parameter interface for this and the ``RosieDiscoService.get`` it overrides.

Originally I left this as work to complete after this PR, but now the urgency has been downgraded due to reviewer preoccupation, I will go & revise the query parameter handling, as above & to include addressing the points in your comments, as that seems sensible with more time.

Reply:
> ...would raise the appropriate error if an invalid parameter if provided. Does Tornado let us do that?
>
> For example would ?formmat=json return an error.

Sorry, I did not address this part in my previous comment, which I deemed & marked as resolving the conversation. In short, given the nature of URL query handling established as standard in Tornado (see https://github.com/metomi/rose/pull/2288#issuecomment-481802596), there is not an easy way to deal with invalid query parameters, so I will leave this as follow-on work, & accordingly I am just about to add it as a new point for the follow-up Issue (text as in https://github.com/metomi/rose/pull/2288#issuecomment-476096016).